### PR TITLE
update run modal description with broadcast info

### PR DESCRIPTION
**Pull request by trilopin**
#### Summary
Screenshot 2022-04-28 at 14 08 37

From https://github.com/mattermost/mattermost-plugin-playbooks/pull/1155

Brand new description in update run status modal.
This PR covers the new description we want to set in the modal for Run "Post update". The new description informs the user about how many channels and DM will be broadcasted with the update. A tooltip will be shown if the user does hover on the counters. Tooltips show up to 5 names and "and N others" in case we have more than 5.

If there are no channels or followers (DM), a different message will be shown in the description.

From the pure tech perspective, a new hook has been implemented to extract the names of the users from the userIDs. Thanks @calebroseland for your help :)

Additionally, the navigation is now done in the same tab (we force modal close if the user navigates). In a different PR, a mechanism to prevent user from losing unsaved changes will be implemented. https://github.com/mattermost/mattermost-plugin-playbooks/pull/1148 has to be merge first.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-43130

#### Checklist
- [ ] ~~Telemetry updated~~
- [ ] ~~Gated by experimental feature flag~~
- [ ] ~~Unit tests updated~~
- [x] E2E tests

### CODE REVIEW

**Comment by trilopin** [comment]
@agarciamontoro @gbochora Sorry for the new PR.

Suggestions from the dirty PR have been addressed, and additionally:
- No broadcast to 0 channels subtitle
- followers are grey underlined instead of blue, don't navigate to run overview, just show tooltip
- respect run toggle to inform about broadcast channels 
- prevent the user from losing changes if accidentally navigates
- navigation with modal opened prepared but commented until @agarciamontoro merges

File: webapp/src/hooks/general.ts [hunk 1]
```diff
@@ -454,6 +454,18 @@ export function useFormattedUsernameByID(userId: string) {
     return useFormattedUsername(user);
 }
 
+// Return the list of names of the users given a list of UserProfiles or userIds
+// It will repect teamnameNameDisplaySetting.
+export function useFormattedUsernames(usersOrUserIds?: Array<UserProfile | string>) : string[] {
+    const teamnameNameDisplaySetting = useSelector<GlobalState, string | undefined>(
```
Comment (gbochora) [inline webapp/src/hooks/general.ts:460]: Should this variable name be `teammateNameDisplaySetting`?

File: webapp/src/hooks/general.ts [hunk 2]
```diff
@@ -454,6 +454,18 @@ export function useFormattedUsernameByID(userId: string) {
     return useFormattedUsername(user);
 }
 
+// Return the list of names of the users given a list of UserProfiles or userIds
+// It will repect teamnameNameDisplaySetting.
```
Comment (gbochora) [inline webapp/src/hooks/general.ts:458]: Did you mean `respect`?

Comment (trilopin) [reply webapp/src/hooks/general.ts:458]: haha yes, fat fingers

File: webapp/src/components/modals/update_run_status_modal.tsx [hunk 3]
```diff
@@ -150,33 +173,56 @@ const UpdateRunStatusModal = ({
...
+                </Tooltip>
+            ),
+            hasFollowersAndChannels: Boolean(broadcastChannelCount && followersChannelCount).toString(),
+            hasChannels: Boolean(broadcastChannelCount).toString(),
+            hasFollowers: Boolean(followersChannelCount).toString(),
+            broadcastChannelCount,
+            followersChannelCount,
+        });
+    };
+
```
Comment (gbochora) [inline webapp/src/components/modals/update_run_status_modal.tsx:222]: Great job here, it's pretty complex.

**Review by gbochora — APPROVED** [review]
Great work @trilopin! 🥳 
One minor thing: when a status update is broadcasted to the private channels I don't have access to, the tooltip message is a little confusing. We could add hidden channels number to the `others`, i.e. `featureX and 5 others` or `IncidentY and 2 others`. 
In any case, I think it's not that important, so I'll leave it up to you.

Screenshot 2022-05-04 at 14 38 08

File: webapp/src/hooks/general.ts (hunk 1 above)
Comment (trilopin) [reply webapp/src/hooks/general.ts:460]: maybe, it's more accurate 👍🏽

**Comment by trilopin** [comment]
> Great work @trilopin! 🥳 One minor thing: when a status update is broadcasted to the private channels I don't have access to, the tooltip message is a little confusing. We could add hidden channels number to the `others`, i.e. `featureX and 5 others` or `IncidentY and 2 others`. In any case, I think it's not that important, so I'll leave it up to you.
> 
> Screenshot 2022-05-04 at 14 38 08

nice feedback @gbochora you rock 🪨

File: tests-e2e/cypress/integration/channels/rhs/status_update_spec.js [hunk 4]
```diff
@@ -241,7 +242,39 @@ describe('channels > rhs > status update', () => {
...
+            it.only('click overview link, go back and save', () => {
+                // # Run the `/playbook update` slash command.
+                cy.executeSlashCommand('/playbook update');
+
+                // # Get dialog modal.
+                cy.getStatusUpdateDialog().within(() => {
+                    // # Type the invalid data
+                    cy.findByTestId('update_run_status_textbox').clear().type('My valid and important changes that I don\'t want to lose');
+
+                    // * Click overview link
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/channels/rhs/status_update_spec.js:254]: Nit:
```suggestion
                    // # Click overview link
```

File: tests-e2e/cypress/integration/channels/rhs/status_update_spec.js [hunk 5]
```diff
@@ -241,7 +242,39 @@ describe('channels > rhs > status update', () => {
...
+                    cy.findByTestId('run-overview-link').click();
+                });
+
+                // * Go back from unsaved changes modal
+                cy.get('#confirm-modal-light').within(() => {
+                    cy.findByTestId('modal-cancel-button').click();
+                });
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/channels/rhs/status_update_spec.js:261]: This does test that the modal is open, but what about some specific check, as in:

```suggestion
                // * Verify that the confirmation modal is shown
                cy.get('#confirm-modal-light').should('exist').within(() => {
                    // # Go back from unsaved changes modal
                    cy.findByTestId('modal-cancel-button').click();
                });
```

File: tests-e2e/cypress/integration/channels/rhs/status_update_spec.js [hunk 6]
```diff
@@ -241,7 +242,39 @@ describe('channels > rhs > status update', () => {
...
+                cy.getStatusUpdateDialog().within(() => {
+                    cy.get('button.confirm').click();
+                });
+
+                // * Verify that the Post update and unsaved changes modals have gone.
+                cy.getStatusUpdateDialog().should('not.exist');
+                cy.get('#confirm-modal-light').should('not.exist');
+            });
+
+            it('cancel and discard explicitily', () => {
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/channels/rhs/status_update_spec.js:277]: Typo
```suggestion
            it('cancel and discard explicitly', () => {
```

File: tests-e2e/cypress/integration/channels/rhs/status_update_spec.js [hunk 7]
```diff
@@ -263,6 +296,29 @@ describe('channels > rhs > status update', () => {
                 cy.getStatusUpdateDialog().should('not.exist');
                 cy.get('#confirm-modal-light').should('not.exist');
             });
+
+            it('click overview link and discard explicitily', () => {
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/channels/rhs/status_update_spec.js:300]: Typo
```suggestion
            it('click overview link and discard explicitly', () => {
```

File: tests-e2e/cypress/integration/channels/rhs/status_update_spec.js [hunk 8]
```diff
@@ -263,6 +296,29 @@ describe('channels > rhs > status update', () => {
...
+                // # Run the `/playbook update` slash command.
+                cy.executeSlashCommand('/playbook update');
+
+                // # Get dialog modal.
+                cy.getStatusUpdateDialog().within(() => {
+                    // # Type the invalid data
+                    cy.findByTestId('update_run_status_textbox').clear().type('My valid and important changes that I don\'t want to lose');
+
+                    // * Click overview link
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/channels/rhs/status_update_spec.js:309]: Nit:
```suggestion
                    // # Click overview link
```

File: tests-e2e/cypress/integration/channels/rhs/status_update_spec.js [hunk 9]
```diff
@@ -263,6 +296,29 @@ describe('channels > rhs > status update', () => {
...
+                    cy.findByTestId('run-overview-link').click();
+                });
+
+                // * Discard explicitily from unsaved changes
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/channels/rhs/status_update_spec.js:313]: Typo :smile: 
```suggestion
                // * Discard explicitly from unsaved changes
```

**Review by agarciamontoro — APPROVED** [review]
Looks great, @trilopin, thank you! :)

I've left only nits and typos, so approving already!

File: tests-e2e/cypress/integration/channels/rhs/status_update_spec.js (hunk 9 above)
Comment (trilopin) [reply tests-e2e/cypress/integration/channels/rhs/status_update_spec.js]: done, Thank you!

**Comment by trilopin** [comment]
@gbochora I implemented your suggestion, thank you!

@abhijit-singh Please take a look. It's everything ready except the navigation with the modal opened because I'm waiting for another merge.

Behavior:
- tooltips: we get as many names as possible (channels could be hidden or unknown) and show up to 5 + "and N others" 
- 0 channels broadcast text removed
- navigation only from broadcast channels, not followers
- if there is no broadcast nor followers, we just paste the overview link

**Review by abhijit-singh — CHANGES_REQUESTED** [review]
Thanks, @trilopin! Looks awesome.

A couple of requests: 
1. The `You have unsaved changes` modal seems to overlay on top of the `Post update` modal and both are visible. This is not consistent with how we've been handling multi-modal flows before. When the second modal is open, the first one should not be visible. Just like it works for the `Confirm finish run` modal. Attaching gifs for both cases here.
`Current`:
`Ideal behaviour`:

 2. It seems like if you hit `Cancel` or `X` on the "You have unsaved changes" modal, and then try to `Cancel` posting the status update, you get stuck in a loop. 


 3. This is minor and may already get solved with the first two points above, but it seems like the `N channels` tooltip gets triggered automatically when you hit `Cancel` on the "You have unsaved changes" modal, and does not go away until you click somewhere else. 

4. When we only show `This update will be saved to overview page`, can we add a full stop `.` at the end? This'll make it consistent with other cases.

**Comment by trilopin** [comment]
> 1. The `You have unsaved changes` modal seems to overlay on top of the `Post update` modal and both are visible. This is not consistent with how we've been handling multi-modal flows before. When the second modal is open, the first one should not be visible. Just like it works for the `Confirm finish run` modal. Attaching gifs for both cases here.

I missed this and it was always there :). It's fixed now (not pushed yet).


> 2. It seems like if you hit `Cancel` or `X` on the "You have unsaved changes" modal, and then try to `Cancel` posting the status update, you get stuck in a loop.

When you are in an unsaved modal you have two options:
- go back to edit: triggered by any type of cancel (X, click outside or cancel button)
- dismiss changes and continue

So if you cancel the unsaved, go back and if you cancel again you go again to unsaved-changes modal. In the end we ask the user to do an explicit dismiss or save. 

does it feel weird? What would you expect to happen? @abhijit-singh 

> 3. This is minor and may already get solved with the first two points above, but it seems like the `N channels` tooltip gets triggered automatically when you hit `Cancel` on the "You have unsaved changes" modal, and does not go away until you click somewhere else.

Wow, let me check this detail

> 4. When we only show `This update will be saved to overview page`, can we add a full stop `.` at the end? This'll make it consistent with other cases.

Thanks for the detailed check (done but not pushed yet)

**Comment by abhijit-singh** [comment]
> So if you cancel the unsaved, go back and if you cancel again you go again to unsaved-changes modal. In the end we ask the user to do an explicit dismiss or save.

@trilopin You're right. I missed the fact that we're showing a `You have unsaved changes` modal if you click on `Cancel` or `X` on the Post Update modal after typing something. Sorry for the confusion here, you can ignore point 2 in my review, that's working as expected.

**Comment by trilopin** [comment]
> This is minor and may already get solved with the first two points above, but it seems like the N channels tooltip gets triggered automatically when you hit Cancel on the "You have unsaved changes" modal, and does not go away until you click somewhere else.

It seems fixed after properly hiding the main modal when the unsaved one is triggered.

**Review by abhijit-singh — APPROVED** [review]
Perfect 💯!

**Pull request merged**